### PR TITLE
Roll up of improvements made during threading work

### PR DESCRIPTION
**Pull request by benvanik**

### CODE REVIEW

File: iree/base/threading_pthreads.c [hunk 1]
```diff
@@ -270,6 +274,23 @@ void iree_thread_override_end(iree_thread_override_t* override) {
...
+void iree_thread_request_affinity(iree_thread_t* thread,
+                                  iree_thread_affinity_t affinity) {
+  if (!affinity.specified) return;
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  // NOTE: Android uses Linux lightweight processes (LWP) for threads, so the
+  // pid is really the tid.
+  pid_t tid = gettid();
+
+  struct cpu_set_t cpu_set;
```
Comment (ScottTodd) [inline iree/base/threading_pthreads.c:286]: Errors on CI ([cmake logs](https://source.cloud.google.com/results/invocations/241fde0a-ec13-4cbe-84d0-e3a1d1c03dda/targets/iree%2Fgcp_ubuntu%2Fcmake%2Flinux%2Fx86-swiftshader%2Fpresubmit/log), [bazel logs](https://source.cloud.google.com/results/invocations/e3843019-a908-4378-bf23-a59c9e59f279/targets/iree%2Fgcp_ubuntu%2Fbazel%2Flinux%2Fx86-swiftshader%2Fcore%2Fpresubmit/log)):
```
../iree/base/threading_pthreads.c:286:20: error: variable has incomplete type 'struct cpu_set_t'
  struct cpu_set_t cpu_set;
                   ^
```


Also some warnings:
```
../iree/base/threading_pthreads.c:24:9: warning: '_GNU_SOURCE' macro redefined [-Wmacro-redefined]
#define _GNU_SOURCE
        ^
<command line>:2:9: note: previous definition is here
#define _GNU_SOURCE 1
        ^
../iree/base/threading_pthreads.c:284:15: warning: implicit declaration of function 'gettid' is invalid in C99 [-Wimplicit-function-declaration]
  pid_t tid = gettid();
              ^
../iree/base/threading_pthreads.c:286:10: note: forward declaration of 'struct cpu_set_t'
  struct cpu_set_t cpu_set;
         ^
```

**Review by ScottTodd — COMMENTED** [review]
Thanks for organizing these into individual commits and splitting them off from your other work :)

File: iree/base/tracing.h [hunk 2]
```diff
@@ -266,6 +266,15 @@ enum {
...
 #define IREE_TRACE_SET_THREAD_NAME(name) iree_tracing_set_thread_name_impl(name)
 
+// Evalutes the expression code only if tracing is enabled.
+//
+// Example:
+//  struct {
+//    IREE_TRACE(uint32_t trace_only_value);
+//  } my_object;
+//  IREE_TRACE(my_object.trace_only_value = 5);
+#define IREE_TRACE(expr) expr
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/base/tracing.h:276]: The intent was to match LLVM_DEBUG.

`*_IF_*` is a tricky word as this is a simple expression passthrough and does not behave like a proper statement on its own (as with macros like RETURN_IF_ERROR do).

**Comment by benvanik** [comment]
Fixed _GNU_SOURCE (*shakes fist* GNUUUUUUUUUUUUU); struct cpu_set_t may require some other trickery I'll look at tomorrow - then I think this is ready.

**Review by ScottTodd — APPROVED** [review]

**Pull request merged**